### DISCUSSION TITLE
Navigate not working for some reason

### DISCUSSION THREAD

**Discussion by damilola1ade** [remix-run/remix#6342] [Q&A]
I'm experiencing a weird scenario. I have a condition whereby a user will be navigated to an onboarding page if he hasn't created his organization. After completing and trying to dispatch the form, it doesn't navigate to the intended route("/") but just resets the form remains on the onboarding page. I can see the organization has been created in my ORM but for some reason the user remains on the onboarding page. What could I be doing wrong please?

This is how I'm handling the redirect.

```jsx
const { message } = useSelector(
  (state) => state.auth
);

const navigate = useNavigate()

useEffect(() => {
  if (message === "This user does not have a business account, please create one and continue.") {
    navigate("/auth/onboarding");
  } else {
    if (message === "success")
    navigate('/')
  }
}, [])
```

This is my onboarding page.

```jsx
const Onboarding = () => {

  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { organization, message, isError, isSuccess, isLoading } = useSelector(
    (state) => state.organization
  );

  useEffect(() => {
    if (isError) {
      toast({
        title: "Error",
        description: message,
        status: "error",
        position: "top-right",
        duration: 5000,
        isClosable: true,
      });
    }

    if (isSuccess) {
      toast({
        title: "Congrats!🎉",
        description: "Onboarding completed.",
        status: "success",
        position: "top-right",
        duration: 5000,
        isClosable: true,
      });
      setTimeout(() => {
        navigate("/admin/dashboard"), 5000;
      });
    }

    dispatch(reset());
  }, [organization, isSuccess, isError]);
```

![Screenshot 2023-05-07 141550](https://user-images.githubusercontent.com/76736466/236679801-6d302f1c-d64c-441a-bc4a-18a7352e7859.png)
As you can the organization creates successfully, but what doesn't it navigate me to ("/")?

**Comment by hakimLyon** [comment]
Hi ! 
Based on ur code, it seems that the redirection to the home page ```"/"``` only occurs when message has the value of "success". However, in your onboarding page, you call ```dispatch(reset())``` every time the component updates. Therefore, it's possible that message is reset before the redirection condition is ever fulfilled.

I suggest moving the call to```dispatch(reset())``` into a specific cleanup function, rather than in the dependency array of your ```useEffect()```. That way, it will only be called once when the component unmounts.

Additionally, I notice that you're using a ```setTimeout()``` function to redirect to the dashboard page after ```5 s``` when ```isSuccess``` is true. This can cause issues if the user leaves the onboarding page before the redirection occurs. Instead, u should use a ```useEffect()``` with ```isSuccess``` as a dependency to trigger the redirection immediately after a successful onboarding.

Here's the corrected code for ur redirect function:

```ts
const { message } = useSelector((state) => state.auth);
const navigate = useNavigate();

useEffect(() => {
  if (message === "This user does not have a business account, please create one and continue.") {
    navigate("/auth/onboarding");
  } else if (message === "success") {
    navigate('/');
  }
}, [message, navigate]);

```
And here's the corrected code for ur onboarding page:
```ts
const Onboarding = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { organization, message, isError, isSuccess } = useSelector((state) => state.organization);

  useEffect(() => {
    if (isError) {
      toast({
        title: "Error",
        description: message,
        status: "error",
        position: "top-right",
        duration: 5000,
        isClosable: true,
      });
    }

    if (isSuccess) {
      toast({
        title: "Congrats!🎉",
        description: "Onboarding completed.",
        status: "success",
        position: "top-right",
        duration: 5000,
        isClosable: true,
      });
    }

    return () => {
      dispatch(reset());
    }
  }, [dispatch, isError, isSuccess, message]);

  useEffect(() => {
    if (isSuccess) {
      navigate("/admin/dashboard");
    }
  }, [isSuccess, navigate]);

  // ...
}
```

🤞 with these modifications, the redirection should work correctly after the organization is created.

**Reply by hakimLyon** [reply]
Also, as an alternative solution, you could consider using the ```useFetcher()```. This hook provides an easy way to make fetch requests and handle their responses, and it might simplify your code.

You can find more information on how to use this hook in the official documentation at https://remix.run/docs/en/1.16.0/hooks/use-fetcher.

**Comment by machour** [comment]
You have an error in your setTimeout call:

instead of 

```jsx
    setTimeout(() => {
        navigate("/admin/dashboard"), 5000;
      });
```

do

```jsx
    setTimeout(() => {
        navigate("/admin/dashboard");
      }, 5000);
```
